### PR TITLE
feat: add `AwsS3EntityProvider` as replacement for `AwsS3DiscoveryProcessor`

### PR DESCRIPTION
**Pull request by pjungermann**
Add a new provider `AwsS3EntityProvider` as a replacement for the now deprecated
`AwsS3DiscoveryProcessor`.

The new provider will scan configured S3 buckets (with optional) prefix and
add `Location` entities for all discovered catalog files.

These `Location` entities will then be processed as usual.

At each execution, the provider will apply a full mutation, replacing all previous
entities with the new entities/state.

Relates-to: #10183
Signed-off-by: Patrick Jungermann <Patrick.Jungermann@gmail.com>

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: .changeset/friendly-hairs-happen.md
Comment (jhaals) [inline .changeset/friendly-hairs-happen.md:5]: Big ⭐ for providing a helpful migration guide.

File: .changeset/friendly-hairs-happen.md
Comment (jhaals) [inline .changeset/friendly-hairs-happen.md:36]: Hmm, I'm not sure about adding a new entityProvider and placing the configuration under `catalog.processors.awsS3.providers`. Perhaps we can do better? @backstage/maintainers suggestions?

File: plugins/catalog-backend-module-aws/api-report.md
Comment (jhaals) [inline plugins/catalog-backend-module-aws/api-report.md:71]: I think most of these can be made private right?

File: plugins/catalog-backend-module-aws/api-report.md
Comment (jhaals) [inline plugins/catalog-backend-module-aws/api-report.md:60]: I'd say make this constructor private and replace it with a `create` method which accepts an options object.

**Review by jhaals — COMMENTED** [review]
Nice work Patrick! 🙌 I've added some suggestions :)

**Comment by pjungermann** [comment]
I rebased on master since PR #10454 was merged.

File: .changeset/friendly-hairs-happen.md
Comment (pjungermann) [reply .changeset/friendly-hairs-happen.md:36]: I tried to follow the example of existing providers like ldapOrg or microsoftGraphOrg. But I get what you mean. Open for suggestions :)

File: plugins/catalog-backend-module-aws/api-report.md
Comment (pjungermann) [reply plugins/catalog-backend-module-aws/api-report.md:60]: currently, it is only used as part of `AwsS3EntityProviderFactory.fromConfig(...)`. I thought about putting the logic from it into it, too, but wanted to keep it cleaner and separated. As the factory uses it as return type I had to export it, too.

But I see your point. The options object is easier extensible. I will add it.

File: plugins/catalog-backend-module-aws/api-report.md
Comment (pjungermann) [reply plugins/catalog-backend-module-aws/api-report.md:71]: Yes. I was wondering which of them need to be private and which of them to keep open for overriding, etc. I guess `protected` does not really make a big difference for what your after either.

I can mark them private and however needs anything could make a PR to open it up more again :)

Comment (pjungermann) [reply plugins/catalog-backend-module-aws/api-report.md:71]: done

File: plugins/catalog-backend-module-aws/api-report.md
Comment (pjungermann) [reply plugins/catalog-backend-module-aws/api-report.md:60]: done

File: .changeset/friendly-hairs-happen.md
Comment (pjungermann) [reply .changeset/friendly-hairs-happen.md:36]: maybe an alternative which could be applied here and for the upcoming processor to provider conversion as of #10183 

```yaml
catalog:
  providers:
    awsS3:
      - bucket: bucket-name
        # [...]
      - bucket: another-bucket-name
        # [...]
```

I guess it would be great to migrate existing providers to it, too, then.

File: plugins/catalog-backend-module-aws/src/credentials/AwsCredentials.ts
Comment (jhaals) [inline plugins/catalog-backend-module-aws/src/credentials/AwsCredentials.ts:25]: `create` instead of build to follow the naming convention a bit more :)

File: .changeset/friendly-hairs-happen.md
Comment (pjungermann) [reply .changeset/friendly-hairs-happen.md:36]: values like bucket names are usually environment-specific. Doing it "hardcoded"/in code would require me still to put it in my own config and load it from there instead.

And if I think about existing providers, this approach might be tricky for them.

But I'm generally more in favor of using the config file.

Comment (pjungermann) [reply .changeset/friendly-hairs-happen.md:36]: seems like we only need alignment/a decision here to move on :)

Comment (pjungermann) [reply .changeset/friendly-hairs-happen.md:36]: I will opt-in for `catalog.providers.awsS3` for now. Let me know what you think

Comment (pjungermann) [reply .changeset/friendly-hairs-happen.md:36]: btw: we also cases like `catalog.processors.awsOrganization.provider` used for the **processor** `AwsOrganizationCloudAccountProcessor`

Comment (Rugvip) [reply .changeset/friendly-hairs-happen.md:36]: The main thing I'm worried about is supporting multiple instances of the provider, which can get tricky depending on how we structure the config. Both for this specific provider and as a general pattern. Totally agree that if this is often environment specific then there should be some built-in way to have that information in config. Ideas welcome 😅

Comment (pjungermann) [reply .changeset/friendly-hairs-happen.md:36]: the current setup assumes multiple instances by default (config array) and the factory returns an array of providers as well

Comment (pjungermann) [reply .changeset/friendly-hairs-happen.md:36]: Of course, one provider instance could handle all buckets (+ prefixes), too, and iterate over them/the config array. But multiple instances can help to spread the load is what I've considered here for this specific provider.

There might be cases where a single instance of a provider is enough. As a general pattern, supporting multiple seems still better. But this does not need to be limited, too. `catalog.providers.{name}` as config prefix could be enough unification/requirement. We could expect providers to always support multiple instances (the value is a config array with the config per instance) or just leave it open how it is structured (single, multiple, multiple with some basic config, etc.).

Comment (Rugvip) [reply .changeset/friendly-hairs-happen.md:36]: To clarify I'm thinking multiple providers of the same type but with different options and configuration. Think for example 3 different buckets with different scheduling configuration, or custom transformation functions. It's not something that existing providers really support and I don't think we have to support it in this PR, but I would like to find a pattern that works well with that so curious if you have any thoughts.

**Comment by benjdlambert** [comment]
Hmm just wondering if we should skip the deprecation of the older stuff for now, just until we can confirm that the provider API's that we've settled on are right and do everything that we want before losing support for things that might have been supported in the `Processor` approach? Does that make sense for now at least?

**Comment by pjungermann** [comment]
> Hmm just wondering if we should skip the deprecation of the older stuff for now, just until we can confirm that the provider API's that we've settled on are right and do everything that we want before losing support for things that might have been supported in the `Processor` approach? Does that make sense for now at least?

@benjdlambert the provider does the same thing as the processor -- just that changes will not cause orphans. Also, deprecating it will not remove it yet. This breaking change could come at a later step.

But of course, deprecating it suggests users to migrate.

At the moment, I don't see a value of the processor anymore though. But maybe I miss something.

If you prefer to remove the deprecation it can be done, of course. Not sure about the docs then though. I would find it a bit confusing to have both there (as user).

**Comment by benjdlambert** [comment]
Yeah sorry, I didn't mean in your implementation, I just meant from our side let's not get people to migrate until we're settled on `Providers` for the catalog. Maybe you could break the documentation up into Recommended and Other? @freben thoughts here on what we should be encouraging?

**Comment by pjungermann** [comment]
@benjdlambert as a user, I would prefer providers due to the orphans created when using processors.

Also a reason why I contributed this one and plan to contribute a few more.

As next level, I would see event-based updates vs those (full) syncs. A combination like initial full sync + event-based updates + full syncs once in a while as auto-healing would work fine.

For event-based updates, webhook-based change events as from SCM systems or other evens like S3 events can be interesting, maybe buffered in a queue; so general JMS support might make sense. Inside backstage, this could be solved with adaptors for the different providers converting events into a unified event schema or triggering APIs directly.
I know there are a couple of thoughts floating around for this and all these need to go through an RFC first.
Just for context.

File: docs/integrations/aws-s3/discovery.md
Comment (pjungermann) [inline docs/integrations/aws-s3/discovery.md:6]: I was wondering whether this should be renamed from Discovery to Ingestion 🤔

File: .changeset/friendly-hairs-happen.md
Comment (pjungermann) [reply .changeset/friendly-hairs-happen.md:36]: With a shared base interface including scheduling options some of this could even be moved up the chain a bit and force some unification -- if desired :)

Comment (freben) [reply .changeset/friendly-hairs-happen.md:36]: Hm. I guess it can be argued that the config here is an `AwsS3DiscoveryProviderFactory` concern, not an `AwsS3EntityProvider` concern. So it would actually be up to the adopter whether they want to / need to make a non-blessed / esoteric setup where they manually instantiate providers, or whether they leverage the fast-lane factory helper. So ... from THAT angle maybe this is entirely OK.

File: plugins/catalog-backend-module-aws/src/providers/AwsS3EntityProviderFactory.ts
Comment (freben) [inline plugins/catalog-backend-module-aws/src/providers/AwsS3EntityProviderFactory.ts:34]: the signature of our `fromConfig` methods are always `(config: Config, options: { ...everything else })`.

File: docs/integrations/aws-s3/discovery.md
Comment (freben) [reply docs/integrations/aws-s3/discovery.md:6]: hm this type of info page for all integrations are (historically) named discovery, so let's change that separately some other time

File: plugins/catalog-backend-module-aws/config.d.ts
Comment (freben) [inline plugins/catalog-backend-module-aws/config.d.ts:45]: initial reaction here:

- is the region necessary? usually the s3 client can deduce it right? unless we HAVE to go via urls.
- i wonder how this relates to the other places that we have config for s3, such as [here](https://github.com/backstage/backstage/blob/949243d991709364647ccdd3acf9625a06179754/plugins/techdocs-backend/config.d.ts#L62) and [here](https://github.com/backstage/backstage/blob/949243d991709364647ccdd3acf9625a06179754/packages/integration/config.d.ts#L210).

It's getting pretty spread out, wondering if there should be ONE place that you configure the actual integration (probably under integrations), and then everything else just "references" those. Maybe that's actually kind of what happens here, but the region parameter threw me off a bit. Also the choice of naming patterns, "bucket" vs "bucketName" etc

**Review by freben — COMMENTED** [review]
Man, it's getting way too late, some early comments before reading through fully

File: plugins/catalog-backend-module-aws/config.d.ts
Comment (pjungermann) [reply plugins/catalog-backend-module-aws/config.d.ts:45]: for S3 bucket, is is required to provide a region as they are region specific and not necessarily in the same region as you are. We could make it optional though and use the default region (e.g., the deployment's region).

File: plugins/catalog-backend-module-aws/src/providers/AwsS3EntityProviderFactory.ts
Comment (pjungermann) [reply plugins/catalog-backend-module-aws/src/providers/AwsS3EntityProviderFactory.ts:34]: 👍  I will adjust it

File: plugins/catalog-backend-module-aws/config.d.ts
Comment (pjungermann) [reply plugins/catalog-backend-module-aws/config.d.ts:45]: re bucket vs bucketName: I'm open for both. At other places, bucket is used (e.g., AwsS3UrlReader.ts) or Bucket at the S3 APIs, but there is also bucketName at techdocs.

I can change it to bucketName.

File: plugins/catalog-backend-module-aws/src/providers/AwsS3EntityProvider.ts
Comment (pjungermann) [reply plugins/catalog-backend-module-aws/src/providers/AwsS3EntityProvider.ts:149]: readTree loads all objects, yes.

I thought downloading all objects and keeping the streams and having the buffers for them (even if concurrency is limited) it might be a bit heavy depending on the size of the bucket.
See also `AwsS3DiscoveryProcessor.doRead`.

And then we still need to execute something like the `CatalogProcessorParser` on top to convert the yaml file content into actual entities, etc.

With the approach of using Locations,
- we can spread the load across the cluster
- can refresh individual ones via the UI
- don't need to "reimplement" the logic by AwsS3UrlReader and following processors

The logic is pretty light-weight and could be applied in a similar way for SCM services (e.g., the provider for Bitbucket Cloud, GitHub, etc.).

File: .changeset/friendly-hairs-happen.md
Comment (pjungermann) [reply .changeset/friendly-hairs-happen.md:7]: The deprecation will still allow to continue using it until they get really removed eventually, but of course, it still encourages to switch.

Personally, I don't see a benefit of the existing processor over the provider.

But that is your strategy and call. I can move this out into a separate PR (or drop it for now) and create a separate documentation instead of the modification to discovery.md, too.

Let me know.

File: plugins/catalog-backend-module-aws/config.d.ts
Comment (pjungermann) [reply plugins/catalog-backend-module-aws/config.d.ts:45]: I made region optional. It will use the default region as fallback.
And I renamed `bucket` -> `bucketName`.

Comment (pjungermann) [reply plugins/catalog-backend-module-aws/config.d.ts]: And regarding integrations: It uses the credentials set up under integrations.

File: .changeset/friendly-hairs-happen.md
Comment (pjungermann) [reply .changeset/friendly-hairs-happen.md:36]: What do you think about adding
```
schedule:
  frequency: PT30M # or cron
  timeout: PT3M
  initialDelay: PT5M # optional
```
to the config read by the factory?

File: .changeset/friendly-hairs-happen.md
Comment (freben) [reply .changeset/friendly-hairs-happen.md:7]: I'd say: don't deprecate yet, but do update docs to highlight the provider as the primary path to take, demoting the processor.

The processor lets you more easily dynamically drive it through the use of locations, which arguably isn't a great use case in and of itself, but it's familiar and people use it out there so it's still "valid" even though it has some distinct drawbacks...

File: .changeset/friendly-hairs-happen.md
Comment (freben) [reply .changeset/friendly-hairs-happen.md:36]: Hold off on that. Good idea, but if we'd go that path, we would want it to be 100% powered by backend-tasks and having some form of fromConfig in there to read that config in a centralised way. Some other day, let's get this one in first

File: plugins/catalog-backend-module-aws/src/providers/AwsS3EntityProvider.ts
Comment (saan800) [inline plugins/catalog-backend-module-aws/src/providers/AwsS3EntityProvider.ts:74]: Could we have `credentials: AwsCredentials` as an optional `create` and `constructor` argument, then default to this create method?

Would allow more flexibility

Comment (pjungermann) [reply plugins/catalog-backend-module-aws/src/providers/AwsS3EntityProvider.ts:74]: Can you sketch a use case where the use of integration would not be sufficient enough?

Comment (pjungermann) [reply plugins/catalog-backend-module-aws/src/providers/AwsS3EntityProvider.ts:74]: PS: originally I thought about adding the credential config to this plugin itself. But due to this need to be aligned with the integration due to the use of the AwsS3UrlReader down the road, I got rid of the duplication

File: .changeset/friendly-hairs-happen.md
Comment (pjungermann) [reply .changeset/friendly-hairs-happen.md:36]: alright, I will skip this for now

File: .changeset/friendly-hairs-happen.md
Comment (pjungermann) [reply .changeset/friendly-hairs-happen.md:7]: alright. Then I will remove the deprecation, split the docs (recover the old and put the current one into a new page "Ingestion.md" maybe) and formulate it a bit milder.

PS: this provider also uses Locations. It is pretty similar to the processor. Which is also the case for a few others where we have both. For me as user, it is a bit hard to see why I should prefer the processor over the provider - but maybe it is just me. The argument of provider over processor would be the absence of orphans.

File: plugins/catalog-backend-module-aws/src/providers/AwsS3EntityProvider.ts
Comment (saan800) [reply plugins/catalog-backend-module-aws/src/providers/AwsS3EntityProvider.ts:74]: 1. My company has a bunch of auto setup for things like role names when we deploy components, and they differ per environment. Having a single role name wouldn't work when we have multiple environments deployed to the same aws account.

2. Local dev: when working locally, we setup our aws creds outside of backstage (eg powershell environment), so there's no risk of someone accidentally committing `app-config.yaml` with creds. We roll our AwsCreds to work for both deployed and local dev options.

File: .changeset/friendly-hairs-happen.md
Comment (freben) [reply .changeset/friendly-hairs-happen.md:7]: Just one article (at the original filename) is fine, it'd just have an "oh and by the way you can also use a processor but it's not recommended" section with its own heading at the bottom). Look at the updates I did for https://backstage.io/docs/integrations/ldap/org for inspiration

Comment (pjungermann) [reply .changeset/friendly-hairs-happen.md:7]: done

File: plugins/catalog-backend-module-aws/src/providers/AwsS3EntityProvider.ts
Comment (pjungermann) [reply plugins/catalog-backend-module-aws/src/providers/AwsS3EntityProvider.ts:74]: All of what you describe should work I believe. We also use RBAC or temporary credentials (locally via profiles).

I assume it is about 
```
'backstage-aws-s3-provider'
```
This is only the `roleSessionName`, not a role or anything.

Comment (saan800) [reply plugins/catalog-backend-module-aws/src/providers/AwsS3EntityProvider.ts:74]: Ahh ok, in which case hopefully should be fine :-)

File: plugins/catalog-backend-module-aws/src/providers/AwsS3EntityProviderFactory.ts
Comment (Rugvip) [inline plugins/catalog-backend-module-aws/src/providers/AwsS3EntityProviderFactory.ts:33]: We don't typically have separate factory classes, unless you're actually creating a factory. I think `fromConfig` could go on the implementation class instead

File: plugins/catalog-backend-module-aws/src/providers/AwsS3EntityProvider.ts
Comment (Rugvip) [inline plugins/catalog-backend-module-aws/src/providers/AwsS3EntityProvider.ts:110]: It's very important that this ID is stable. Tbh I don't think we should be coupling it to the bucket name, as could be that you want to switch that out. Thinking it's better to use an explicit ID here instead that's provided by the integrator

File: .changeset/friendly-hairs-happen.md
Comment (Rugvip) [reply .changeset/friendly-hairs-happen.md:36]: I'd say that the transforms are only specific to the `MicrosoftGraphOrgEntityProvider` right now, expecting that pattern to show up in more places tbh.

Potentially something like this?

```yaml
catalog:
  providers:
    awsS3:
      default:
        bucket: bucket-name
```

Along with 

```ts
AwsS3DiscoveryProviderFactory.fromConfig(env.config, { logger: env.logger });
// OR
AwsS3DiscoveryProviderFactory.fromConfig(env.config, { logger: env.logger, id: 'default' });
```

With multiple providers just being more keys with different IDs.

Then another option that avoids having to always add the extra config level:

```yaml
catalog:
  providers:
    awsS3:
      # Not an array, implicit 'default'
      bucket: bucket-name
```

```yaml
catalog:
  providers:
    awsS3:
      # Array format, id is required
    - id: 'default'
      bucket: bucket-name
    - id: 'other'
      bucket: bucket-name
```

Again, just bouncing some ideas :grin: Although see the separate comment about wanting stable IDs, which ends up being a bit related to this

File: plugins/catalog-backend-module-aws/src/providers/AwsS3EntityProvider.ts
Comment (pjungermann) [reply plugins/catalog-backend-module-aws/src/providers/AwsS3EntityProvider.ts:110]: I see. I've considered them sort of stable as they will be stable per config. But you are right, if you want to change this, it may cause problems.

I will solve this with your suggestion for an identifier. 👍

File: .changeset/friendly-hairs-happen.md
Comment (pjungermann) [reply .changeset/friendly-hairs-happen.md:36]: @Rugvip I guess there is the question whether we want to support the implicit ID option (single provider case) or always assume multiple.

For this provider, multiple seems fine to me, but as a general pattern, maybe supporting both is desired and then doing it here would be reasonable (and not hard anyways).

Probably, I would go for first example using named configs instead of the array.

File: plugins/catalog-backend-module-aws/src/providers/AwsS3EntityProviderFactory.ts
Comment (pjungermann) [reply plugins/catalog-backend-module-aws/src/providers/AwsS3EntityProviderFactory.ts:33]: I tried to reduce the dependencies and complexity a bit, but I guess I can find a way for both more or less. Get your point.

File: plugins/catalog-backend-module-aws/src/providers/AwsS3EntityProvider.ts
Comment (pjungermann) [reply plugins/catalog-backend-module-aws/src/providers/AwsS3EntityProvider.ts:150]: @freben Any thoughts on that still or can I assume this is resolved?

Comment (freben) [reply plugins/catalog-backend-module-aws/src/providers/AwsS3EntityProvider.ts:150]: @Rugvip Any opinions on this one?

Comment (freben) [reply plugins/catalog-backend-module-aws/src/providers/AwsS3EntityProvider.ts:150]: I'll resolve this one. I realize that error handling becomes harder to do right too, when just some partial errors in a fraction of the files are found and so on. Unless Rugvip chimes in with something extra, consider this closed

Comment (Rugvip) [reply plugins/catalog-backend-module-aws/src/providers/AwsS3EntityProvider.ts:150]: Yep thinking that all makes sense @pjungermann. I feel we'd need to power up the entity provider scheduling capabilities if we wanted to have the individual reads happen eagerly @freben.

Something to add into the mix could also be a nicer way for providers to invalidate the entity hash so that they get bumped to the front of the queue if needed, without having to add something like a ETag inside the entity.

File: plugins/catalog-backend-module-aws/src/providers/AwsS3EntityProviderFactory.ts
Comment (pjungermann) [reply plugins/catalog-backend-module-aws/src/providers/AwsS3EntityProviderFactory.ts]: I moved it to the provider class and removed the factory.

File: plugins/catalog-backend-module-aws/src/providers/AwsS3EntityProvider.ts
Comment (pjungermann) [reply plugins/catalog-backend-module-aws/src/providers/AwsS3EntityProvider.ts]: I use the "id" as part of the name, now.

File: .changeset/friendly-hairs-happen.md
Comment (pjungermann) [reply .changeset/friendly-hairs-happen.md:36]: I use the following approach now
```
catalog:
  providers:
    awsS3:
      providerId:
        bucket: bucket-name
```

**Comment by pjungermann** [comment]
just added the _option_ to omit the provider ID in the config for the single config use case using "default" as ID.

```yaml
# app-config.yaml

catalog:
  providers:
    awsS3:
      # uses "default" as provider ID
      bucketName: sample-bucket
      prefix: prefix/ # optional
      region: us-east-2 # optional, uses the default region otherwise
```

File: plugins/catalog-backend-module-aws/api-report.md
Comment (pjungermann) [inline plugins/catalog-backend-module-aws/api-report.md:37]: Just realized that this is not needed anymore to be exported since the factory's fromConfig was integrated and the previous create method is gone. The constructor is private, too.

Comment (pjungermann) [reply plugins/catalog-backend-module-aws/api-report.md]: removed

**Comment by pjungermann** [comment]
@Rugvip @freben I believe it is ready to be merged :-)

File: plugins/catalog-backend-module-aws/src/providers/AwsS3EntityProvider.ts
Comment (freben) [inline plugins/catalog-backend-module-aws/src/providers/AwsS3EntityProvider.ts:65]: thought for the future: this actually can never be undefined at the current point in time, but if we ever changed the behavior to not add a default entry silently, it might have been wise to be defensive here and throw an error if integration turned out to be undefined

File: plugins/catalog-backend-module-aws/src/providers/AwsS3EntityProvider.ts
Comment (freben) [inline plugins/catalog-backend-module-aws/src/providers/AwsS3EntityProvider.ts:206]: we would have to check if the target was really an amazon one right?

**Review by freben — APPROVED** [review]
Let's merge and tweak!

File: plugins/catalog-backend-module-aws/src/providers/AwsS3EntityProvider.ts
Comment (pjungermann) [reply plugins/catalog-backend-module-aws/src/providers/AwsS3EntityProvider.ts:65]: good point. I can create a PR to add this.

File: plugins/catalog-backend-module-aws/src/providers/AwsS3EntityProvider.ts
Comment (pjungermann) [reply plugins/catalog-backend-module-aws/src/providers/AwsS3EntityProvider.ts:206]: We already check these cases in the lines before. If we don't point to the real AWS S3, we need to provide the `endpoint` config property (e.g., when LocalStack is used, etc.). So, I'd say it is already covered.

File: plugins/catalog-backend-module-aws/src/providers/AwsS3EntityProvider.ts
Comment (pjungermann) [reply plugins/catalog-backend-module-aws/src/providers/AwsS3EntityProvider.ts:65]: follow-up PR #10697

**Comment by pjungermann** [comment]
I just realized that applying the "allow" catalog rule to locations emitted by an entity provider only supports the global config. 

For the awsS3 processor, this was not really the case as the processor's Location can have a the Location-specific "allow" rule and itself only emits the entities contained within the S3 objects/catalog files. Hence, the rule was applied successfully.

For other processors which emit Locations as well like the SCM-related processors for github, gitlab, bitbucket, etc. should be impacted by this limitation as well.

In order to mitigate this, you need to ensure that to be added entities are allowed by the global rule.

Not yet sure what would be the right strategy to have a custom configuration for the provider (instance) and be more restrictive at the global rule. 🤔

**Comment by pjungermann** [comment]
issue #12880 as follow up related to the "allow" rule

**Pull request merged**
